Looks like we'll need to cfg `gungraun` out on WASM:

```
error[E0277]: the trait bound `OsString: serde::Serialize` is not satisfied
    --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gungraun-runner-0.19.2/src/api.rs:1110:44
     |
1110 |   #[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
     |                                              ^^^^^^^^^ the trait `Serialize` is not implemented for `OsString`
...
1118 | /     /// The environment variables to set or pass through to the binary
1119 | |     pub envs: Vec<(OsString, Option<OsString>)>,
     | |_______________________________________________- required by a bound introduced by this call
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -203,6 +203,25 @@ version = "1.0.52"
 [dev-dependencies.rustversion]
 version = "1"
 
+[target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies.gungraun]
+version = "=0.19.2"
+
+[[bench]]
+name = "gungraun_format_str"
+harness = false
+
+[[bench]]
+name = "gungraun_parse_str"
+harness = false
+
+[[bench]]
+name = "gungraun_v4"
+harness = false
+
+[[bench]]
+name = "gungraun_v7"
+harness = false
+
 [workspace]
 members = [
     "rng",

--- a/benches/gungraun_format_str.rs
+++ b/benches/gungraun_format_str.rs
@@ -1,0 +1,83 @@
+// Gungraun benchmarks require the `gungraun-runner` binary and Valgrind, which
+// aren't available on wasm32 targets, so the harness is compiled out there.
+#[cfg(not(target_arch = "wasm32"))]
+mod bench {
+    use std::hint::black_box;
+    use std::io::Write;
+
+    use gungraun::prelude::*;
+    use uuid::Uuid;
+
+    fn setup_uuid() -> Uuid {
+        Uuid::parse_str("F9168C5E-CEB2-4faa-B6BF-329BF39FA1E4").unwrap()
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_uuid())]
+    fn hyphenated(uuid: Uuid) -> [u8; 36] {
+        let mut buffer = [0_u8; 36];
+        write!(&mut buffer as &mut [_], "{:x}", uuid.hyphenated()).unwrap();
+        black_box(buffer)
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_uuid())]
+    fn simple(uuid: Uuid) -> [u8; 32] {
+        let mut buffer = [0_u8; 32];
+        write!(&mut buffer as &mut [_], "{:x}", uuid.simple()).unwrap();
+        black_box(buffer)
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_uuid())]
+    fn urn(uuid: Uuid) -> [u8; 36 + 9] {
+        let mut buffer = [0_u8; 36 + 9];
+        write!(&mut buffer as &mut [_], "{:x}", uuid.urn()).unwrap();
+        black_box(buffer)
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_uuid())]
+    fn encode_hyphen(uuid: Uuid) -> [u8; 36] {
+        let mut buffer = [0_u8; 36];
+        uuid.hyphenated().encode_lower(&mut buffer);
+        black_box(buffer)
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_uuid())]
+    fn encode_simple(uuid: Uuid) -> [u8; 32] {
+        let mut buffer = [0_u8; 32];
+        uuid.simple().encode_lower(&mut buffer);
+        black_box(buffer)
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_uuid())]
+    fn encode_urn(uuid: Uuid) -> [u8; 36 + 9] {
+        let mut buffer = [0_u8; 36 + 9];
+        uuid.urn().encode_lower(&mut buffer);
+        black_box(buffer)
+    }
+
+    library_benchmark_group!(
+        name = format_str,
+        benchmarks = [
+            hyphenated,
+            simple,
+            urn,
+            encode_hyphen,
+            encode_simple,
+            encode_urn
+        ]
+    );
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+use bench::format_str;
+
+#[cfg(not(target_arch = "wasm32"))]
+gungraun::main!(library_benchmark_groups = format_str);
+
+#[cfg(target_arch = "wasm32")]
+fn main() {}

--- a/benches/gungraun_parse_str.rs
+++ b/benches/gungraun_parse_str.rs
@@ -1,0 +1,88 @@
+// Gungraun benchmarks require the `gungraun-runner` binary and Valgrind, which
+// aren't available on wasm32 targets, so the harness is compiled out there.
+#[cfg(not(target_arch = "wasm32"))]
+mod bench {
+    use std::hint::black_box;
+
+    use gungraun::prelude::*;
+    use uuid::Uuid;
+
+    #[library_benchmark]
+    fn parse_nil() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box("00000000000000000000000000000000")))
+    }
+
+    #[library_benchmark]
+    fn parse_nil_hyphenated() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box(
+            "00000000-0000-0000-0000-000000000000",
+        )))
+    }
+
+    #[library_benchmark]
+    fn parse_random() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box("67e5504410b1426f9247bb680e5fe0c8")))
+    }
+
+    #[library_benchmark]
+    fn parse_random_hyphenated() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box(
+            "67e55044-10b1-426f-9247-bb680e5fe0c8",
+        )))
+    }
+
+    #[library_benchmark]
+    fn parse_urn() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box(
+            "urn:uuid:67e55044-10b1-426f-9247-bb680e5fe0c8",
+        )))
+    }
+
+    #[library_benchmark]
+    fn parse_invalid_len() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box("F9168C5E-CEB2-4faa-BBF-329BF39FA1E4")))
+    }
+
+    #[library_benchmark]
+    fn parse_invalid_character() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box(
+            "F9168C5E-CEB2-4faa-BGBF-329BF39FA1E4",
+        )))
+    }
+
+    #[library_benchmark]
+    fn parse_invalid_group_len() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box("01020304-1112-2122-3132-41424344")))
+    }
+
+    #[library_benchmark]
+    fn parse_invalid_groups() -> Result<Uuid, uuid::Error> {
+        black_box(Uuid::parse_str(black_box(
+            "F9168C5E-CEB2-4faa-B6BFF329BF39FA1E4",
+        )))
+    }
+
+    library_benchmark_group!(
+        name = parse_str,
+        benchmarks = [
+            parse_nil,
+            parse_nil_hyphenated,
+            parse_random,
+            parse_random_hyphenated,
+            parse_urn,
+            parse_invalid_len,
+            parse_invalid_character,
+            parse_invalid_group_len,
+            parse_invalid_groups
+        ]
+    );
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+use bench::parse_str;
+
+#[cfg(not(target_arch = "wasm32"))]
+gungraun::main!(library_benchmark_groups = parse_str);
+
+#[cfg(target_arch = "wasm32")]
+fn main() {}

--- a/benches/gungraun_v4.rs
+++ b/benches/gungraun_v4.rs
@@ -1,0 +1,28 @@
+// Gungraun benchmarks require the `gungraun-runner` binary and Valgrind, which
+// aren't available on wasm32 targets, so the harness is compiled out there.
+#[cfg(all(not(target_arch = "wasm32"), feature = "v4"))]
+mod bench {
+    use std::hint::black_box;
+
+    use gungraun::prelude::*;
+    use uuid::Uuid;
+
+    #[library_benchmark]
+    fn new_v4() -> Uuid {
+        black_box(Uuid::new_v4())
+    }
+
+    library_benchmark_group!(
+        name = v4,
+        benchmarks = [new_v4]
+    );
+}
+
+#[cfg(all(not(target_arch = "wasm32"), feature = "v4"))]
+use bench::v4;
+
+#[cfg(all(not(target_arch = "wasm32"), feature = "v4"))]
+gungraun::main!(library_benchmark_groups = v4);
+
+#[cfg(not(all(not(target_arch = "wasm32"), feature = "v4")))]
+fn main() {}

--- a/benches/gungraun_v7.rs
+++ b/benches/gungraun_v7.rs
@@ -1,0 +1,80 @@
+// Gungraun benchmarks require the `gungraun-runner` binary and Valgrind, which
+// aren't available on wasm32 targets, so the harness is compiled out there.
+#[cfg(all(not(target_arch = "wasm32"), feature = "v7", feature = "std"))]
+mod bench {
+    use std::hint::black_box;
+    use std::time::SystemTime;
+
+    use gungraun::prelude::*;
+    use uuid::{ContextV7, NoContext, Timestamp, Uuid};
+
+    #[library_benchmark]
+    fn now_v7() -> Uuid {
+        black_box(Uuid::now_v7())
+    }
+
+    #[library_benchmark]
+    fn new_v7_no_context() -> Uuid {
+        black_box(Uuid::new_v7(Timestamp::now(NoContext)))
+    }
+
+    fn setup_context() -> ContextV7 {
+        ContextV7::new()
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_context())]
+    fn new_v7_context(ctxt: ContextV7) -> Uuid {
+        black_box(Uuid::new_v7(Timestamp::now(&ctxt)))
+    }
+
+    fn setup_context_additional_precision() -> ContextV7 {
+        ContextV7::new().with_additional_precision()
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_context_additional_precision())]
+    fn new_v7_context_additional_precision(ctxt: ContextV7) -> Uuid {
+        black_box(Uuid::new_v7(Timestamp::now(&ctxt)))
+    }
+
+    fn setup_raw() -> (u64, u32) {
+        let now = SystemTime::UNIX_EPOCH.elapsed().unwrap();
+        (now.as_secs(), now.subsec_nanos())
+    }
+
+    #[library_benchmark]
+    #[bench::default(setup_raw())]
+    fn v7_raw((secs, subsec_nanos): (u64, u32)) -> Uuid {
+        let mut counter = 0;
+        black_box(Uuid::new_v7(Timestamp::from_unix_time(
+            secs,
+            subsec_nanos,
+            {
+                counter += 1;
+                counter
+            },
+            42,
+        )))
+    }
+
+    library_benchmark_group!(
+        name = v7,
+        benchmarks = [
+            now_v7,
+            new_v7_no_context,
+            new_v7_context,
+            new_v7_context_additional_precision,
+            v7_raw
+        ]
+    );
+}
+
+#[cfg(all(not(target_arch = "wasm32"), feature = "v7", feature = "std"))]
+use bench::v7;
+
+#[cfg(all(not(target_arch = "wasm32"), feature = "v7", feature = "std"))]
+gungraun::main!(library_benchmark_groups = v7);
+
+#[cfg(not(all(not(target_arch = "wasm32"), feature = "v7", feature = "std")))]
+fn main() {}

--- a/src/fmt.rs
+++ b/src/fmt.rs
@@ -179,22 +179,22 @@ impl Uuid {
     }
 }
 
-const UPPER: [u8; 16] = [
-    b'0', b'1', b'2', b'3', b'4', b'5', b'6', b'7', b'8', b'9', b'A', b'B', b'C', b'D', b'E', b'F',
-];
-const LOWER: [u8; 16] = [
-    b'0', b'1', b'2', b'3', b'4', b'5', b'6', b'7', b'8', b'9', b'a', b'b', b'c', b'd', b'e', b'f',
-];
+// Maps a hex nibble (0..=15) to its ASCII character. `alpha_offset` is added
+// for values above 9: 0x27 for lowercase, 0x07 for uppercase.
+#[inline]
+const fn nibble_to_hex(nibble: u8, alpha_offset: u8) -> u8 {
+    nibble + b'0' + if nibble > 9 { alpha_offset } else { 0 }
+}
 
 #[inline]
 const fn format_simple(src: &[u8; 16], upper: bool) -> [u8; 32] {
-    let lut = if upper { &UPPER } else { &LOWER };
+    let alpha_offset = if upper { 0x07 } else { 0x27 };
     let mut dst = [0; 32];
     let mut i = 0;
     while i < 16 {
         let x = src[i];
-        dst[i * 2] = lut[(x >> 4) as usize];
-        dst[i * 2 + 1] = lut[(x & 0x0f) as usize];
+        dst[i * 2] = nibble_to_hex(x >> 4, alpha_offset);
+        dst[i * 2 + 1] = nibble_to_hex(x & 0x0f, alpha_offset);
         i += 1;
     }
     dst
@@ -202,27 +202,33 @@ const fn format_simple(src: &[u8; 16], upper: bool) -> [u8; 32] {
 
 #[inline]
 const fn format_hyphenated(src: &[u8; 16], upper: bool) -> [u8; 36] {
-    let lut = if upper { &UPPER } else { &LOWER };
-    let groups = [(0, 8), (9, 13), (14, 18), (19, 23), (24, 36)];
+    let simple = format_simple(src, upper);
     let mut dst = [0; 36];
 
-    let mut group_idx = 0;
     let mut i = 0;
-    while group_idx < 5 {
-        let (start, end) = groups[group_idx];
-        let mut j = start;
-        while j < end {
-            let x = src[i];
-            i += 1;
-
-            dst[j] = lut[(x >> 4) as usize];
-            dst[j + 1] = lut[(x & 0x0f) as usize];
-            j += 2;
-        }
-        if group_idx < 4 {
-            dst[end] = b'-';
-        }
-        group_idx += 1;
+    while i < 8 {
+        dst[i] = simple[i];
+        i += 1;
+    }
+    dst[8] = b'-';
+    while i < 12 {
+        dst[i + 1] = simple[i];
+        i += 1;
+    }
+    dst[13] = b'-';
+    while i < 16 {
+        dst[i + 2] = simple[i];
+        i += 1;
+    }
+    dst[18] = b'-';
+    while i < 20 {
+        dst[i + 3] = simple[i];
+        i += 1;
+    }
+    dst[23] = b'-';
+    while i < 32 {
+        dst[i + 4] = simple[i];
+        i += 1;
     }
     dst
 }

--- a/src/parser.rs
+++ b/src/parser.rs
@@ -202,35 +202,26 @@ pub(crate) const fn parse_simple(
         ));
     }
 
-    let mut buf: [u8; 16] = [0; 16];
+    // Copy the hex characters into a fixed-size array so the decoder's
+    // bounds are statically known.
+    let mut hex = [0u8; 32];
     let mut i = 0;
-
-    while i < 16 {
-        // Convert a two-char hex value (like `A8`)
-        // into a byte (like `10101000`)
-        let h1 = HEX_TABLE[s[i * 2] as usize];
-        let h2 = HEX_TABLE[s[i * 2 + 1] as usize];
-
-        // We use `0xff` as a sentinel value to indicate
-        // an invalid hex character sequence (like the letter `G`)
-        if h1 | h2 == 0xff {
-            return Err(InvalidUuid(
-                s,
-                if speculative {
-                    RequestedUuid::Any
-                } else {
-                    RequestedUuid::Simple
-                },
-            ));
-        }
-
-        // The upper nibble needs to be shifted into position
-        // to produce the final byte value
-        buf[i] = SHL4_TABLE[h1 as usize] | h2;
+    while i < 32 {
+        hex[i] = s[i];
         i += 1;
     }
 
-    Ok(buf)
+    match decode_hex32(&hex) {
+        Some(buf) => Ok(buf),
+        None => Err(InvalidUuid(
+            s,
+            if speculative {
+                RequestedUuid::Any
+            } else {
+                RequestedUuid::Simple
+            },
+        )),
+    }
 }
 
 #[inline]
@@ -241,8 +232,7 @@ pub(crate) const fn parse_hyphenated(s: &'_ [u8]) -> Result<[u8; 16], InvalidUui
         return Err(InvalidUuid(s, RequestedUuid::Hyphenated));
     }
 
-    // We look at two hex-encoded values (4 chars) at a time because
-    // that's the size of the smallest group in a hyphenated UUID.
+    // The hex characters are split into five groups separated by hyphens.
     // The indexes we're interested in are:
     //
     // uuid     : 936da01f-9abd-4d9d-80c7-02af85c822a8
@@ -256,66 +246,80 @@ pub(crate) const fn parse_hyphenated(s: &'_ [u8]) -> Result<[u8; 16], InvalidUui
         _ => return Err(InvalidUuid(s, RequestedUuid::Hyphenated)),
     }
 
-    let positions: [u8; 8] = [0, 4, 9, 14, 19, 24, 28, 32];
-    let mut buf: [u8; 16] = [0; 16];
-    let mut j = 0;
-
-    while j < 8 {
-        let i = positions[j];
-
-        // The decoding here is the same as the simple case
-        // We're just dealing with two values instead of one
-        let h1 = HEX_TABLE[s[i as usize] as usize];
-        let h2 = HEX_TABLE[s[(i + 1) as usize] as usize];
-        let h3 = HEX_TABLE[s[(i + 2) as usize] as usize];
-        let h4 = HEX_TABLE[s[(i + 3) as usize] as usize];
-
-        if h1 | h2 | h3 | h4 == 0xff {
-            return Err(InvalidUuid(s, RequestedUuid::Hyphenated));
-        }
-
-        buf[j * 2] = SHL4_TABLE[h1 as usize] | h2;
-        buf[j * 2 + 1] = SHL4_TABLE[h3 as usize] | h4;
-        j += 1;
+    // Gather the hex characters, skipping the four hyphens, so they're
+    // contiguous for the decoder.
+    let mut hex = [0u8; 32];
+    let mut i = 0;
+    while i < 8 {
+        hex[i] = s[i];
+        i += 1;
+    }
+    while i < 12 {
+        hex[i] = s[i + 1];
+        i += 1;
+    }
+    while i < 16 {
+        hex[i] = s[i + 2];
+        i += 1;
+    }
+    while i < 20 {
+        hex[i] = s[i + 3];
+        i += 1;
+    }
+    while i < 32 {
+        hex[i] = s[i + 4];
+        i += 1;
     }
 
-    Ok(buf)
+    match decode_hex32(&hex) {
+        Some(buf) => Ok(buf),
+        None => Err(InvalidUuid(s, RequestedUuid::Hyphenated)),
+    }
 }
 
-const HEX_TABLE: &[u8; 256] = &{
-    let mut buf = [0; 256];
-    let mut i: u8 = 0;
+/// Decodes 32 hexadecimal ASCII characters into 16 bytes, returning `None` if
+/// any character is not a valid hexadecimal digit.
+#[inline]
+const fn decode_hex32(hex: &[u8; 32]) -> Option<[u8; 16]> {
+    let mut nibbles = [0u8; 32];
+    let mut bad = 0u8;
 
-    loop {
-        buf[i as usize] = match i {
-            b'0'..=b'9' => i - b'0',
-            b'a'..=b'f' => i - b'a' + 10,
-            b'A'..=b'F' => i - b'A' + 10,
-            _ => 0xff,
+    let mut i = 0;
+    while i < 32 {
+        let c = hex[i];
+
+        // '0'..='9' map to 0..=9; 'a'..='f' and 'A'..='F' (via `| 0x20`) map to
+        // 0..=5, offset by 10 to land in 10..=15.
+        let digit = c.wrapping_sub(b'0');
+        let alpha = (c | 0x20).wrapping_sub(b'a');
+
+        let is_digit = digit < 10;
+        let is_alpha = alpha < 6;
+
+        nibbles[i] = if is_digit {
+            digit
+        } else {
+            alpha.wrapping_add(10)
         };
 
-        if i == 255 {
-            break buf;
-        }
-
-        i += 1
-    }
-};
-
-const SHL4_TABLE: &[u8; 256] = &{
-    let mut buf = [0; 256];
-    let mut i: u8 = 0;
-
-    loop {
-        buf[i as usize] = i.wrapping_shl(4);
-
-        if i == 255 {
-            break buf;
-        }
+        bad |= if is_digit | is_alpha { 0 } else { 1 };
 
         i += 1;
     }
-};
+
+    if bad != 0 {
+        return None;
+    }
+
+    let mut buf = [0u8; 16];
+    let mut j = 0;
+    while j < 16 {
+        buf[j] = (nibbles[j * 2] << 4) | nibbles[j * 2 + 1];
+        j += 1;
+    }
+
+    Some(buf)
+}
 
 #[cfg(test)]
 mod tests {

